File: demos/typescript/src/customer-service.ts
export interface Customer {
  id: string;
  name: string;
  email: string;
  isActive: boolean;
}

// Beroenden uttrycks som interfaces, precis som i .NET-demot.
export interface CustomerRepository {
  getById(id: string): Promise<Customer | null>;
  save(customer: Customer): Promise<void>;
}

export interface EmailSender {
  send(to: string, subject: string, body: string): Promise<void>;
}

export class CustomerService {
  // Beroenden in via konstruktorn -> lätt att skicka in mockar i test.
  constructor(
    private readonly repo: CustomerRepository,
    private readonly email: EmailSender
  ) {}

  async deactivate(id: string): Promise<boolean> {
    const customer = await this.repo.getById(id);
    if (!customer) return false; // okänd kund -> gör inget mer

    customer.isActive = false;
    await this.repo.save(customer);
    await this.email.send(
      customer.email,
      "Account deactivated",
      `Hi ${customer.name}, your account has been deactivated.`
    );
    return true;
  }
}

File: demos/dotnet/CustomerApi/Services/CustomerService.cs
using CustomerApi.Models;

namespace CustomerApi.Services;

// Klassisk "orchestrator": hämtar, ändrar, sparar, notifierar.
// Beroenden tas in via konstruktorn (dependency injection) så att
// vi kan skicka in mockar i testerna istället för riktig DB och SMTP.
public class CustomerService
{
    private readonly ICustomerRepository _repo;
    private readonly IEmailSender _email;

    public CustomerService(ICustomerRepository repo, IEmailSender email)
    {
        _repo = repo;
        _email = email;
    }

    public async Task<bool> DeactivateAsync(Guid id)
    {
        var customer = await _repo.GetByIdAsync(id);
        if (customer is null)
            return false; // Okänd kund -> gör inget mer, skicka inget mail.

        customer.IsActive = false;
        await _repo.SaveAsync(customer);
        await _email.SendAsync(
            customer.Email,
            "Account deactivated",
            $"Hi {customer.Name}, your account has been deactivated.");
        return true;
    }
}
